sycl/CMakeLists.txt

```diff
@@ -137,12 +137,17 @@ COMMENT "Copying SYCL headers ...")
 # Configure SYCL headers
 install(DIRECTORY "${sycl_inc_dir}/." DESTINATION "${LLVM_INST_INC_DIRECTORY}" COMPONENT sycl-headers)
 
+set(SYCL_RT_LIBS sycl)
```

Comment:
I don't think we should build two versions of SYCL runtime library. I consider this patch as a work-around for poor SYCL library ABI design.
We didn't pay much attention to the library ABI design and it .
I would prefer you to review binary interface and make it stable and portable instead of this change.

Reply:
@bader - Alexey, Do you suggest replacing all standard classes such as vector, etc with something else? 
Those debug|release versions of C Runtime Libraries are linked when you use any of this impressive list of C++ header files: https://docs.microsoft.com/en-us/cpp/standard-library/cpp-standard-library-header-files?view=vs-2019

See, it is written here: https://docs.microsoft.com/en-us/cpp/c-runtime-library/crt-library-features?view=vs-2019

> When you build a release version of your project, one of the basic C run-time libraries (libcmt.lib, msvcmrt.lib, msvcrt.lib) is linked by default, depending on the compiler option you choose (multithreaded, DLL, /clr). If you include one of the C++ Standard Library header files in your code, a C++ Standard Library will be linked in automatically by Visual C++ at compile time.

Reply:
Adding to my previous statement.
I believe the poor ABI is not caused by our library implementation, but by the SYCL SPEC.
SYCL SPEC makes us using those std::vector, std::string, etc
Here is extract from it:

> namespace cl {
9 namespace sycl {
10
11 template < class T, class Alloc = std::allocator<T> >
12 using vector_class = std::vector<T, Alloc>;
13
14 using string_class = std::string;
15
16 template<class R, class... ArgTypes>
17 using function_class = std::function<R(ArgTypes...)>;
18
19 using mutex_class = std::mutex;
20
21 template <class T>
22 using shared_ptr_class = std::shared_ptr<T>;
23
24 template <class T>
25 using unique_ptr_class = std::unique_ptr<T>;
26
27 template <class T>
28 using weak_ptr_class = std::weak_ptr<T>;
29
30 template <class T>
31 using hash_class = std::hash<T>;
32
33 using exception_ptr_class = std::exception_ptr;
34
35 } // sycl
36 } // cl

Reply:
> Alexey, Do you suggest replacing all standard classes such as vector, etc with something else?

Yes.


> SYCL SPEC makes us using those std::vector, std::string, etc

No, it's not. It requires SYCL implementation to define <...>_class classes, which ***can*** be aliases to standard library classes, but specification does not require this particular implementation.

Anyway we do not have to use SYCL API in the library interface. It's an implementation decision - how functionality is split between library and headers and interface is used between them. Using C++ classes in ABI is not recommended as C++ classes layout is implementation defined, so using different compilers to build headers and a library doesn't work in general case.

Reply:
Ok, I understood what you mean, and agree now that it would be great to avoid using std::vector args/returns of exported functions in sycl/source. That would let us have only 1 sycl.dll. Thank you.